File: Leetcode/C++/167. Two Sum II - Input Array Is Sorted/solution.cpp
// https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/submissions/980465431/

#include <vector>

using namespace std;

class Solution // Two pointer: O(N)
{
public:
    vector<int> twoSum(vector<int> &numbers, int target)
    {
        int left = 0, right = numbers.size() - 1;
        int total;
        while (left < right)
        {
            total = numbers[left] + numbers[right];
            if (total > target)
            {
                right--;
            }
            else if (total < target)
            {
                left++;
            }
            else
            {
                // Make 1-index
                left++;
                right++;
                break;
            }
        }
        return {left, right};
    }
};

class Solution // Binary search
{
public:
    vector<int> twoSum(vector<int> &numbers, int target)
    {
        int left, right, mid, num;
        for (int i = 0; i < numbers.size(); i++)
        {
            num = numbers[i];
            left = i + 1;
            right = numbers.size() - 1;
            while (left <= right)
            {
                mid = left + (right - left) / 2;
                if (num + numbers[mid] > target)
                {
                    right = mid - 1;
                }
                else if (num + numbers[mid] < target)
                {
                    left = mid + 1;
                }
                else
                {
                    // Make 1-index and return answer
                    return {i + 1, mid + 1};
                }
            }
        }
        return {-1, -1}; // Dummy data problem guarantee there is one answer
    }
};

File: Leetcode/Python/167. Two Sum II - Input Array Is Sorted/solution.py
# https://leetcode.com/problems/two-sum-ii-input-array-is-sorted/

class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        start = 0
        end = len(numbers) - 1
        while start <= end:
            total = numbers[start] + numbers[end]
            if total > target:
                end -= 1
            elif total < target:
                start += 1
            else:
                return [start + 1, end + 1]


class Solution:
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        dic = {}
        for idx, num in enumerate(numbers):
            if target - num in dic:
                return [dic[target - num] + 1, idx + 1]
            dic[num] = idx

            
class Solution: # Two pointer
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        start, end = 0, len(numbers) - 1
        while start < end:
            if numbers[start] + numbers[end] == target:
                return [start + 1, end + 1]
            elif numbers[start] + numbers[end] > target:
                end -= 1
            else:
                start += 1


class Solution:  # Binary search
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        for idx, val in enumerate(numbers):
            start, end = idx + 1, len(numbers) - 1
            temp_target = target - val
            while start <= end:
                mid = start + (end - start) // 2
                if numbers[mid] == temp_target:
                    return [idx + 1, mid + 1]
                elif numbers[mid] > temp_target:
                    end = mid - 1
                else:
                    start = mid + 1
            